musicApp/R/functions.R

```diff
@@ -167,3 +177,115 @@ getAlbumLyrics = function(albumname, artist, progressBar = FALSE)
+# currently the playlist is made on my own spotify profile (https://open.spotify.com/user/1117862156?si=OMhP-ANnSae9ijUFoqT_ew)
+# because i havent found out how i can let someone without a developper account log in.
+# if you want to try the function you need a playlist ID, which you can get by sharing the link to a playlist
+# (id is the letters/numbers in the url between 'playlist/' and '?si=') note that the playlist you use has to be public
+
+makePlaylist <- function(playlistID){
+
+  # create dataframe with data about the playlist and create variable with list of artists.
+  playlist = get_playlist_audio_features('Karel Veldkamp', playlistID)
+  artists = playlist %>%
+    unnest(track.artists) %>%
+    select(name, id) %>%
+    unique
+
+
+  # sample 30 artists (if n > 30)
+  n = nrow(artists)
+  if (n > 30)
+  {
+    sample <- sample(1:n, 30)
+  }else
+  {
+    sample <- 1:n
+  }
+
+  # create a pool of artist that have collaborated with the artists in the playlist, or that are related to them.
+  artistPool = data.frame()
+  for (i in sample)
+  {
+    collabs = get_artist_albums(artists[i, 'id']) %>%
+      select(artists) %>%
+      unnest(artists) %>%
+      filter(!name %in% artists$name) %>%
+      select(id) %>%
+      unique()
+
+    related = get_related_artists(artists[i, 'id'])[1:5,] %>%
+      filter(!name %in% artists$name) %>%
+      select(id)
+
+    artistPool = rbind(artistPool, collabs, related)
+  }
+
+  # get top tracks from each artist in the artist pool
+  allTracks = data.frame()
+  for (i in 1:nrow(artistPool))
+  {
+    toptracks <- get_artist_top_tracks(artistPool[i,])
+    if (length(toptracks)>0){
+      if (! toptracks$id %in% allTracks$id){
+        tracks <- get_track_audio_features(toptracks$id)
+        allTracks <- rbind(allTracks, tracks)
+      }
+    }
+  }
+
+  # get matrix with scaled audio features.
+  audioFeatures = as.matrix(allTracks[features])
+  audioFeatures = scale(audioFeatures)
+  rownames(audioFeatures) <- allTracks$uri
+
+  # to pick which songs are most similar to the ones in the user playlist. I use k means clustering:
+
+  # first save the audio features of songs in user playlist (scaled to a z score)
+  features = c('danceability', 'energy', 'loudness', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence')
+  playlistProfile = scale(playlist[features])
+
+
+  # then calculate the mean silhouette coefficient for differnt numbers of k to decide the optimal k value:
+  scores = 0  # first score is set to 0 because sillouette coefficient is not defined for k == 1
+  maxCluster = min(nrow(playlistProfile)-1,10) # try up to 10 means
+
+  # this loop tries k means clustering with 2 up to 10 clusters:
+  for (k in 2:maxCluster)
+  {
+    kmeans = kmeans(playlistProfile,k)
+    sCoefs = silhouette(kmeans$cluster, dist(playlistProfile))[,3]
+    scores = c(scores, mean(sCoefs))
+  }
+
+  # I then see which k lead to the highest silhouette score and use that model.
```

Comment:
Great idea to look at the silhouette score! In case the clustering results are unsatisfactory, you could consider another method to determine the ideal maximum number of clusters. For instance, you could see what result maximizes the Dunn index. If both methods don't lead to the same result, you could just take the average. This is rather statistical and won't make a huge difference in practice, but might be worth it to look into if you want to continue this project after this course.